sdk/src/client/api/block_builder/input_selection/requirement/mana.rs

```diff
@@ -233,43 +236,41 @@ impl InputSelection {
         Ok(Unlocks::new(blocks)?)
     }
 
-    pub(crate) fn get_inputs_for_mana_balance(&mut self) -> Result<Vec<InputSigningData>, Error> {
-        let (mut selected_mana, required_mana) = self.mana_sums(true)?;
+    pub(crate) fn get_inputs_for_mana_balance(&mut self) -> Result<bool, Error> {
+        let (mut selected_mana, mut required_mana) = self.mana_sums(true)?;
 
         log::debug!("Mana requirement selected mana: {selected_mana}, required mana: {required_mana}");
 
+        let mut added_inputs = false;
         if selected_mana >= required_mana {
             log::debug!("Mana requirement already fulfilled");
-            Ok(Vec::new())
         } else {
-            let mut inputs = Vec::new();
-
+            if !self.allow_additional_input_selection {
+                return Err(Error::AdditionalInputsRequired(Requirement::Mana));
+            }
             // TODO we should do as for the amount and have preferences on which inputs to pick.
             while let Some(input) = self.available_inputs.pop() {
                 selected_mana += self.total_mana(&input)?;
-                inputs.push(input);
+                if let Some(output) = self.select_input(input)? {
+                    required_mana += output.mana();
+                }
+                added_inputs = true;
 
                 if selected_mana >= required_mana {
                     break;
                 }
             }
-            if selected_mana < required_mana {
-                return Err(Error::InsufficientMana {
-                    found: selected_mana,
-                    required: required_mana,
-                });
-            }
-
-            Ok(inputs)
         }
+        Ok(added_inputs)
     }
 
     pub(crate) fn mana_sums(&self, include_remainders: bool) -> Result<(u64, u64), Error> {
-        let required_mana = if include_remainders {
-            self.all_outputs().map(|o| o.mana()).sum::<u64>() + self.remainders.added_mana
-        } else {
-            self.non_remainder_outputs().map(|o| o.mana()).sum::<u64>()
-        } + self.mana_allotments.values().sum::<u64>();
+        let mut required_mana =
+            self.non_remainder_outputs().map(|o| o.mana()).sum::<u64>() + self.mana_allotments.values().sum::<u64>();
+        if include_remainders {
+            required_mana += self.remainder_outputs().map(|o| o.mana()).sum::<u64>() + self.remainders.added_mana;
```

Comment:
`self.remainders.added_mana` is the mana amount we added to existing outputs so we don't need an own remainder output, right? Would be good to have a comment here, hard to understand otherwise